### PR TITLE
[SPARK-19267][SS]Fix a race condition when stopping StateStore

### PR DESCRIPTION
**Pull request by zsxwing**
## What changes were proposed in this pull request?

There is a race condition when stopping StateStore which makes `StateStoreSuite.maintenance` flaky. `StateStore.stop` doesn't wait for the running task to finish, and an out-of-date task may fail `doMaintenance` and cancel the new task. Here is a reproducer: https://github.com/zsxwing/spark/commit/dde1b5b106ba034861cf19e16883cfe181faa6f3

This PR adds MaintenanceTask to eliminate the race condition.

## How was this patch tested?

Jenkins

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala [hunk 1]
```diff
@@ -198,7 +214,7 @@ object StateStore extends Logging {
   private def doMaintenance(): Unit = {
     logDebug("Doing maintenance")
     if (SparkEnv.get == null) {
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala]: There is another potential issue here: if a SparkContext is created after checking `SparkEnv.get == null`, the following `stop` may cancel a new valid task. However, I think that won't happen in practice, so don't fix it.

**Review by tdas — CHANGES_REQUESTED** [review]
I had a more surgical fix in mind when I said that put multiple volatile objects into a class so that we can replace the class completely. Here is what I had in mind. 

https://github.com/apache/spark/compare/master...tdas:state-store-fix?expand=1

What do you think?

**Comment by zsxwing** [comment]
@tdas could you take another look? I fixed some minor issues in your patch.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala [hunk 2]
```diff
@@ -124,12 +125,42 @@ object StateStore extends Logging {
...
   val MAINTENANCE_INTERVAL_DEFAULT_SECS = 60
 
+  @GuardedBy("loadedProviders")
   private val loadedProviders = new mutable.HashMap[StateStoreId, StateStoreProvider]()
-  private val maintenanceTaskExecutor =
-    ThreadUtils.newDaemonSingleThreadScheduledExecutor("state-store-maintenance-task")
 
-  @volatile private var maintenanceTask: ScheduledFuture[_] = null
-  @volatile private var _coordRef: StateStoreCoordinatorRef = null
+  class MaintenanceTask(periodMs: Long, task: => Unit, onError: => Unit) {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala]: Can you add docs.

Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala]: you should mention the properties of this class. that it automatically cancels the periodic task if there is an exception. and what is onError for.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala [hunk 3]
```diff
@@ -179,14 +210,14 @@ object StateStore extends Logging {
   /** Start the periodic maintenance task if not already started and if Spark active */
   private def startMaintenanceIfNeeded(): Unit = loadedProviders.synchronized {
     val env = SparkEnv.get
-    if (maintenanceTask == null && env != null) {
+    if (env != null && (maintenanceTask == null || !maintenanceTask.isRunning)) {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala]: Can you replace this with the method isMaintenanceRunning?

**Review by tdas — COMMENTED** [review]
Mostly LGTM.

**Comment by tdas** [comment]
LGTM, pending tests.

**Pull request closed without merging**